homeassistant/components/onkyo/media_player.py

```diff
@@ -106,462 +203,405 @@
+
+    @callback
+    def async_onkyo_connect_callback(origin):
+        """Receiver (re)connected."""
+        receiver = receivers[origin]
+        _LOGGER.debug("Receiver (re)connected: %s", receiver.name)
+
+        for entity in entities[origin].values():
+            entity.backfill_state()
+
+    def setup_receiver(receiver):
+        KNOWN_HOSTS.append(receiver.host)
+
+        # Store the receiver object and create a dictionary to store its entities.
+        receivers[receiver.host] = receiver
+        entities[receiver.host] = {}
+
+        # Discover what zones are available for the receiver by querying the power.
+        # If we get a response for the specific zone, it means it is available.
+        for zone in ZONES:
+            receiver.query_property(zone, "power")
+
+        # Add the main zone to entities, since it is always active.
+        main_entity = OnkyoMediaPlayer(
+            receiver, sources, "main", max_volume, receiver_max_volume
+        )
+        entities[receiver.host]["main"] = main_entity
+        async_add_entities([main_entity])
+
+    if host is not None and host not in KNOWN_HOSTS:
+        receiver = await pyeiscp.Connection.create(
+            host=host,
+            update_callback=async_onkyo_update_callback,
+            connect_callback=async_onkyo_connect_callback,
+        )
+
+        # The library automatically adds a name and identifier only on discovered hosts,
+        # so manually add them here instead.
+        receiver.name = name
+        receiver.identifier = None
+
+        setup_receiver(receiver)
     else:
-        for receiver in eISCP.discover():
+
+        @callback
+        async def async_onkyo_discovery_callback(receiver):
+            """Receiver discovered, connection not yet active."""
+            _LOGGER.debug("Receiver discovered: %s (%s)", receiver.name, receiver.host)
             if receiver.host not in KNOWN_HOSTS:
-                hosts.append(OnkyoDevice(receiver, config.get(CONF_SOURCES)))
-                KNOWN_HOSTS.append(receiver.host)
-    add_entities(hosts, True)
+                await receiver.connect()
+                setup_receiver(receiver)
 
+        await pyeiscp.Connection.discover(
+            update_callback=async_onkyo_update_callback,
+            connect_callback=async_onkyo_connect_callback,
+            discovery_callback=async_onkyo_discovery_callback,
+        )
+
+    @callback
+    def close_receiver(_event):
+        for receiver in receivers.values():
+            receiver.close()
+
+    hass.bus.async_listen_once(EVENT_HOMEASSISTANT_STOP, close_receiver)
 
-class OnkyoDevice(MediaPlayerEntity):
-    """Representation of an Onkyo device."""
 
-    _attr_supported_features = SUPPORT_ONKYO
+class OnkyoMediaPlayer(MediaPlayerEntity):
+    """Representation of an Onkyo Receiver Media Player (one per each zone)."""
+
+    _attr_should_poll = False
+    _attr_sound_mode_list = list(SOUND_MODE_MAPPING)
```

Comment:
This wasn't here before right?

Reply:
Yes, I added sound mode support as well.

Reply:
Let's leave that for followups